Source/Peripheral_Libs/Displays.Waveshare-ePaper/Driver/Drivers/EPD2i7.cs

```diff
@@ -0,0 +1,317 @@
+
+            if (buffer != null)
+            {
+                SendCommand(PARTIAL_DATA_START_TRANSMISSION_2);
+                SendData(x >> 8);
+                SendData(x & 0xf8);     // x should be the multiple of 8, the last 3 bit will always be ignored
+                SendData(y >> 8);
+                SendData(y & 0xff);
+                SendData(width >> 8);
+                SendData(width & 0xf8);     // w (width) should be the multiple of 8, the last 3 bit will always be ignored
+                SendData(height >> 8);
+                SendData(height & 0xff);
+                DelayMs(2);
+                for (int i = 0; i < width / 8 * height; i++)
+                {
+                    SendData(buffer[i]);
+                }
+
+                /* send the image data */
+             /*   for (int j = y; j < y + height; j++)
+                {
+                    for (int i = x; i < (x + width) / 8; i++)
+                    {
+                        SendData(buffer[i + j * (width / 8)]);
+                    }
+                }*/
+
+
+
+                DelayMs(2);
+            }
+        }
+
+        void RefreshPartial(int x, int y, int w, int l)
+        {
+            SendCommand(PARTIAL_DISPLAY_REFRESH);
+            SendData(x >> 8);
+            SendData(x & 0xf8);     // x should be the multiple of 8, the last 3 bit will always be ignored
+            SendData(y >> 8);
+            SendData(y & 0xff);
+            SendData(w >> 8);
+            SendData(w & 0xf8);     // w (width) should be the multiple of 8, the last 3 bit will always be ignored
+            SendData(l >> 8);
+            SendData(l & 0xff);
+
+            WaitUntilIdle();
+        }
+
+        void DisplayFrame(byte[] frame_buffer) 
+        {
+            if (frame_buffer == null)
+                return;
+            
+            SendCommand(DATA_START_TRANSMISSION_1);
+            DelayMs(2);
+            for(int i = 0; i< Width / 8 * Height; i++)
+                SendData(0xFF);
+
+            DelayMs(2);
+            SendCommand(DATA_START_TRANSMISSION_2);
+            DelayMs(2);
+
+            SendData(frame_buffer);
+  
+            DelayMs(2);
+            SendCommand(DISPLAY_REFRESH);
+            WaitUntilIdle();
+        }
+
+        public override void DisplayFrame ()
+        {
+            SendCommand(DISPLAY_REFRESH);
+            WaitUntilIdle();
+        }
+
+        public void ClearFrame()
+        {
+            SendCommand(DATA_START_TRANSMISSION_1);
+            DelayMs(2);
+            for (int i = 0; i < Width * Height / 8; i++)
+            {
+                SendData(0xFF);
+            }
+
+            DelayMs(2);
+            SendCommand(DATA_START_TRANSMISSION_2);
+            DelayMs(2);
+
+            for (int i = 0; i < Width * Height / 8; i++)
+            {
+                SendData(0xFF);
+            }
+            DelayMs(2);
+        }
+
+       /* protected override void WaitUntilIdle()
```

Comment:
Want to keep this for safe keeping?

Reply:
It's in the Arduino code so I left it - I agree it feels like it "shouldn't" be needed